[CMake] Support Macros in Linux

rintaro opened the pull request:
For compiling codes required for macro support, we now need swiftc compiler in the build machine.

Unlike Darwin OSes, where swiftCore runtime is guaranteed to be present in /usr/lib, Linux doesn't have ABI stability and the stdlib of the build machine is not at the specific location. So the built compiler cannot relies on the shared object in the system.

rintaro committed:
[CMake] Support Macros in Linux

For compiling codes required for macro support, we now need swiftc
compiler in the build machine.

Unlike Darwin OSes, where swiftCore runtime is guaranteed to be present
in /usr/lib, Linux doesn't have ABI stability and the stdlib of the
build machine is not at the specific location. So the built compiler
cannot relies on the shared object in the toolchain.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented on SwiftCompilerSources/CMakeLists.txt:
```diff
@@ -251,6 +251,12 @@ else()
 
 #include \"Parse/RegexParserBridging.h\"
 ")
+  add_custom_command(
+    OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/HeaderDependencies.cpp"
+    COMMAND ${CMAKE_COMMAND} -E copy_if_different
+      "${CMAKE_CURRENT_BINARY_DIR}/HeaderDependencies.cpp.tmp"
+      "${CMAKE_CURRENT_BINARY_DIR}/HeaderDependencies.cpp"
+  )
```
This is avoid mtime of `HeaderDependencies.cpp` being updated every time `cmake` configuration happens.

rintaro commented on cmake/modules/AddSwift.cmake:
```diff
@@ -548,10 +548,14 @@ function(_add_swift_runtime_link_flags target relpath_to_lib_dir bootstrapping)
+
+      # At runtime, use swiftCore in the current toolchain.
+      # FIXME: This assumes the ABI hasn't changed since the builder.
+      set(swift_runtime_rpath "$ORIGIN/${relpath_to_lib_dir}/swift/${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}")
+
       if(ASRLF_BOOTSTRAPPING_MODE STREQUAL "HOSTTOOLS")
-        set(swift_runtime_rpath "${host_lib_dir}")
-      else()
-        set(swift_runtime_rpath "$ORIGIN/${relpath_to_lib_dir}/swift/${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}")
+        # But before building the stdlib with the tool, use the builder libs. This should be removed in install time.
+        list(APPEND swift_runtime_rpath "${host_lib_dir}")
```
`RUNPATH` is now like `$ORIGIN/../lib/swift/linux:/opt/swift/5.8.1/usr/lib/swift/linux`
So after the `swiftCore` is built in the build directory, that's used.

rintaro commented on cmake/modules/AddSwiftUnittests.cmake:
```diff
@@ -89,7 +111,8 @@ function(add_swift_unittest test_dirname)
 
-  if (SWIFT_SWIFT_PARSER)
+  if (SWIFT_SWIFT_PARSER AND NOT ASU_IS_TARGET_TEST)
+    # Link to stdlib the compiler uses.
```
e.g. `runtime` tests is a "target" test, not "host" test. "target" test should not link with host libraries.

rintaro commented on cmake/modules/SwiftImplicitImport.cmake:
```diff
@@ -6,7 +6,7 @@ function(swift_supports_implicit_module module_name out_var)
       -Xfrontend -disable-implicit-${module_name}-module-import
       -Xfrontend -parse-stdlib
-      -c - -o /dev/null
+      -parse -
```
Linux `swiftc` reads the output file after compiling (for e.g. autolink-extract). Since `/dev/null` is not a valid object file, this failed.
All we want to check here is if the frontend accepts the command line option, so `-parse` should be enough.

rintaro commented on lib/CMakeLists.txt:
```diff
@@ -33,7 +33,7 @@ if (SWIFT_SWIFT_PARSER)
        OUTPUT_VARIABLE SWIFT_SYNTAX_TARGETS)
 
-  set(SWIFT_SYNTAX_LIBRARIES_SOURCE_DIR
+  set(SWIFT_SYNTAX_LIBRARIES_BUILD_DIR
       "${SWIFT_PATH_TO_EARLYSWIFTSYNTAX_BUILD_DIR}/lib/swift/host")
```
This is really meant to be a "build" directory where earlyswiftsyntax's `.so` files exist.

rintaro commented on lib/CMakeLists.txt:
```diff
@@ -53,17 +53,46 @@ if (SWIFT_SWIFT_PARSER)
       COMMENT "Copying ${sharedlib}"
     )
 
+    swift_install_in_component(
+      FILES "${SWIFT_HOST_LIBRARIES_DEST_DIR}/${sharedlib}"
+      DESTINATION "lib${LLVM_LIBDIR_SUFFIX}/swift/host"
+      COMPONENT compiler
+    )
+
+    swift_install_file_set_rpath(
+      "lib${LLVM_LIBDIR_SUFFIX}/swift/host/${sharedlib}"
+      "$$ORIGIN:$$ORIGIN/../linux"
+      compiler
+    )
```
SwiftSyntax libraries (`.so` and `.swiftmodule`) are now being installed from `swift` build directory, but not `earlyswiftsyntax` build directory. So we can centralize the RUNPATH mutation logic in this repository.

rintaro commented on lib/Macros/CMakeLists.txt:
```diff
@@ -31,6 +31,19 @@ function(add_swift_macro_library name)
+  # Add rpath to 'lib/host'
+  if(SWIFT_HOST_VARIANT_SDK STREQUAL "LINUX")
+    set_property(TARGET ${name}
+      APPEND PROPERTY INSTALL_RPATH "$ORIGIN/..")
```
Linux's `RUNPATH` doesn't inherits executable's `RUNPATH`. So each `.so` must have its own complete `RUNPATH`.

finagolfin commented:
I notice you are stripping builder runpaths. I haven't look into why that's needed in detail, but maybe it won't be needed if `-no-toolchain-stdlib-rpath` is passed in so the Swift compiler doesn't set that builder runpath and you set it manually with [CMake's `BUILD_RPATH` instead, which is automatically removed before installation](https://cmake.org/cmake/help/latest/prop_tgt/BUILD_RPATH.html).

That may be easier than removing builder runpaths yourself before installation, as you seem to be doing here? I could be way off since I didn't look deeply into why you're doing this.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

bnbarham commented:
> I notice you are stripping builder runpaths. I haven't look into why that's needed in detail, but maybe it won't be needed if `-no-toolchain-stdlib-rpath` is passed in so the Swift compiler doesn't set that builder runpath and you set it manually with [CMake's `BUILD_RPATH` instead, which is automatically removed before installation](https://cmake.org/cmake/help/latest/prop_tgt/BUILD_RPATH.html).
> 
> That may be easier than removing builder runpaths yourself before installation, as you seem to be doing here? I could be way off since I didn't look deeply into why you're doing this.

The TLDR is that we need to cherry-pick this change to 5.9.0 and thus would like as few changes as possible.

`-no-toolchain-stdlib-rpath` is actually being used for the "pure swift" libs and tools in the compiler, but only so that we can add the host path to the end rather than the start. We could split BUILD/INSTALL_RPATH here rather than using BUILD_WITH_INSTALL_RPATH and then stripping, but that's a much larger change (especially if we want to do it for the non-"pure swift" cases).

We also need to do this for the copied-in "early" swift-syntax libs, which currently aren't part of the swift build itself. Since they're built separately, we can't do the same BUILD/INSTALL_RPATH split. I have a PR up to build swift-syntax in the same tree (https://github.com/apple/swift/pull/66043), but will get back to that after this is in and 5.9 is released (I ran into some issues with the bootstrapping case, which I haven't had the time to look into). I also have a local patch that does most of the BUILD/INSTALL_RPATH splitting that I'll put up as well.

Once that's all in, I definitely agree that using BUILD/INSTALL_RPATH here rather than the manual stripping is the way to go.

rintaro committed:
Avoid adding redundant temporary RUNPATH to builder's stdlib

Many shared libs and executables are only run after stdlib/runtime are
built. They don't need to link with builders stdlib at all.

rintaro committed:
[CMake] Fatal error when bootstrapping is not "hosttools"

and Swift parser integration is enabled.
If swift parser integration is enabled, SwiftSyntax libraries are always
built with host tools. Other SwiftCompilerSources modules must use the
same runtime with parser libraries.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please Build Toolchain Linux Platform

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain Linux Platform

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain macOS Platform

bnbarham commented on tools/driver/CMakeLists.txt:
```diff
@@ -26,6 +26,7 @@ if(BOOTSTRAPPING_MODE MATCHES "BOOTSTRAPPING.*")
     HAS_SWIFT_MODULES
     BOOTSTRAPPING 0
     THINLTO_LD64_ADD_FLTO_CODEGEN_ONLY
+    REQUIRED_FOR_MINIMAL_COMPILER
```
Could skip adding this and the next, since they're only used for bootstrapping which we're not doing.

bnbarham commented on tools/libSwiftScan/CMakeLists.txt:
```diff
@@ -31,13 +31,22 @@ set_target_properties(libSwiftScan
+    # Ensure that we can find the host shared libraries.
+    set_property(
+      TARGET libSwiftScan
+      APPEND PROPERTY INSTALL_RPATH "@loader_path/swift/host")
```
Could just append them both at once with `:` but doesn't really matter.

bnbarham commented on utils/build-presets.ini:
```diff
@@ -859,7 +859,6 @@ install-lldb
 install-llbuild
 install-swiftpm
 install-swift-driver
-install-swiftsyntax
```
Could just leave this in since the produce is changed. There's a lot of these that would need to be removed anyway.

bnbarham commented on utils/build-presets.ini:
```diff
@@ -890,6 +889,8 @@ installable-package=%(installable_package)s
 # in Linux CI bots
 relocate-xdg-cache-home-under-build-subdir
 
+# Linux only support 'hosttools' with early-swiftsyntax.
+bootstrapping=hosttools
```
We're going to need to add this in quite a few places. Maybe we should override and set to hosttools when early swiftsyntax is enabled, since that's the only thing that works? Bit weird but...

rintaro replied:
Updated to force setting `set(BOOTSTRAPPING_MODE "HOSTTOOLS")` when `SWIFT_SWIFT_PARSER` is enabled.

bnbarham commented on utils/swift_build_support/swift_build_support/products/earlyswiftsyntax.py:
```diff
@@ -73,18 +73,10 @@ def test(self, host_target):
-        """
-        self.install_with_cmake(["install"], self.host_install_destdir(host_target))
+        # No-op.
+        None
```
`pass` is the Python-ism you want here

bnbarham commented:
Thanks Rintaro!

> Swift Build Toolchain Ubuntu 20.04 (x86_64) — Done

:O!

rintaro commented on cmake/modules/AddSwift.cmake:
```diff
@@ -744,7 +745,7 @@ function(add_swift_host_library name)
 
   if (ASHL_SHARED AND ASHL_HAS_SWIFT_MODULES)
-      _add_swift_runtime_link_flags(${name} "." "")
+      _add_swift_runtime_link_flags(${name} "." "" FALSE)
```
I don't think any shared (`ASHL_SHARED`) libraries are required for building stdlib.

bnbarham replied:
Ah this is only shared, true. Having said that, if we did move to shared libraries more, it would be the case that they are "required for minimal" but this would still be FALSE because they wouldn't have any Swift. And that was more my point.

rintaro commented on cmake/modules/SwiftStripBuilderRpath.cmake:
```diff
@@ -0,0 +1,71 @@
+  if(NOT(SWIFT_HOST_VARIANT_SDK STREQUAL "LINUX"))
+      return()
+  endif()
+  if((NOT SWIFT_SWIFT_PARSER) AND NOT(BOOTSTRAPPING_MODE STREQUAL "HOSTTOOLS"))
+      return()
+  endif()
```
My intent was  `!(linux && (swift_swift_parser || boostrapping==hosttools))` because `boostrapping==hosttools` alone also needs to strip builder's `RUNPATH`.

bnbarham replied:
`bootstrapping==hosttools` without early swift syntax shouldn't have the host in the `RUNPATH` anyway though right?

rintaro replied:
It should, otherwise which runtime does the `SwiftCompilerSources` library link with (before it builds stdlib/runtime)?

bnbarham replied:
Ah true, hosttools uses SwiftCompilerSources too 😅 Thanks, sorry for the run around!

rintaro committed:
[CMake] Update for review comments

rintaro commented on cmake/modules/AddPureSwift.cmake:
```diff
@@ -65,6 +65,26 @@ function(_add_host_swift_compile_options name)
+
+    get_filename_component(swift_bin_dir ${CMAKE_Swift_COMPILER} DIRECTORY)
+    get_filename_component(swift_dir ${swift_bin_dir} DIRECTORY)
+    set(host_lib_dir "${swift_dir}/lib/swift/${SWIFT_SDK_${SWIFT_HOST_VARIANT_SDK}_LIB_SUBDIR}")
```
They are leftover. Just removed.

rintaro committed:
[CMake] Fix weiredness in swift runtime linking for SourceKit libraries

SourceKit components are never built while bootstrapping stages. So it
doen't need to have stage 0 or 1 handling.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

bnbarham commented on tools/SourceKit/cmake/modules/AddSwiftSourceKit.cmake:
```diff
@@ -129,12 +129,11 @@ function(add_sourcekit_swift_runtime_link_flags target path HAS_SWIFT_MODULES)
+      # So there's no need to add RUNPATH to builder's runtime libraries.
 
     elseif(ASKD_BOOTSTRAPPING_MODE STREQUAL "BOOTSTRAPPING")
       get_bootstrapping_swift_lib_dir(bs_lib_dir "")
```
Ah, (4) adds host. IMO I'd move that to inside the ifs here since the if there is wrong anyway (it should be specifically linux/android/openbsd). So ie.
```
if(SWIFT_SWIFT_PARSER)
    file(RELATIVE_PATH relative_hostlib_path "${path}" "${SWIFTLIB_DIR}/host")
    list(APPEND RPATH_LIST "$ORIGIN/${relative_hostlib_path}")
endif()
```

rintaro replied:
:+1: Done

rintaro committed:
[CMake] Further cleanup for RUNPATH setting

`$ORIGIN` is only supported in Linux, BSD, etc. (i.e. not in Windows)

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro committed:
[CMake] Set LD_LIBRARY_PATH to builder's runtime when building stdlib

In Linux. Instead of setting temporary "fallback" RUNPATH, Set
LD_LIBRARY_PATH to builder's runtime when building standard library.
So we don't need to strip the temporary RUNPATH when installing.

rintaro committed:
[CMake] Fix install for macOS toolchain

recursive-lipo requires shared libraries have +x permissions.
Use PROGRAMS instead of FILES when installing swift-syntax libraries.

DougGregor commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090

@swift-ci Please smoke test

DougGregor commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro committed:
[CMake] Fix RUNPATH of lib_InternalSwiftScan in Linux

libSwiftScan is built in 'lib' but installed in 'lib/swift/host' RUNPATH
should have correct '$ORIGIN/../{platform}' to load 'swiftCore' runtime
library.

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090

@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro committed:
[CMake] Stop some exectuable link with swiftCore

swift-compatibility-symbols, swift-def-to-strings-converter,
and swift-serialize-diagnostics don't use any Swift modules. But when
SWIFT_SWIFT_PARSER was enabled, they are linked with swiftCore. But
these binaries can be executed before the runtime is being built.
We need to stop them linking with swiftCore.

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/llvm-project/pull/7331
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain Windows platform

rintaro committed:
[CMake] Use RPATH_CHANGE instead RPATH_SET

RPATH_SET is not available until cmake 3.21.0. Use RPATH_CHANGE instead.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro committed:
[CMake] Always use builder's runtime for building Swift code in Linux

Some non-stdlib thing e.g. swift-backtrace still might be built before
the runtime is built. For building Swift code in Linux "hosttools",
always set 'LD_LIBRARY_PATH' to the runtime in the builder.

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please Build Toolchain

rintaro commented:
https://github.com/apple/swift-syntax/pull/2090
@swift-ci Please smoke test

finagolfin commented:
I see a handful of runpath issues with [the last CI-built toolchain for linux](https://download.swift.org/tmp/pull-request/68082/553/ubuntu2004/PR-ubuntu2004.tar.gz) after this pull, I'll look into the fixes:
```
> ./usr/bin/lldb-server
./usr/bin/lldb-server: error while loading shared libraries: libswiftCore.so: cannot open shared object file: No such file or directory
> readelf -d ./usr/bin/lldb-server |ag runpath
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN/../lib:$ORIGIN/swift/linux:$ORIGIN/../lib/swift/host]
```
That second relative path is wrong, hence the runtime error.

bnbarham commented:
Looks like https://github.com/apple/llvm-project/blob/63a082e326d053e778a05a899514ba101d7db5f6/lldb/cmake/modules/AddLLDB.cmake#L225 needs a `${reldir}lib`

finagolfin commented on tools/libSwiftScan/CMakeLists.txt:
```diff
@@ -30,14 +30,36 @@ set_target_properties(libSwiftScan
+  elseif(SWIFT_HOST_VARIANT_SDK MATCHES "LINUX|ANDROID|OPENBSD|FREEBSD")
+    set_property(
+      TARGET libSwiftScan
+      APPEND PROPERTY INSTALL_RPATH "$ORIGIN/swift/host")
```
@rintaro, is this needed for some intermediate build step or something? Because the final installed library also has this path, which is wrong:
```
>

bnbarham replied:
Yeah, see the many FIXME above. This all needs to be split into BUILD/INSTALL_RPATH + built into host rather than into lib, but that's too much for 5.9. We'll fix this up after all the 5.9 PRs are in as well. We could remove it, but it doesn't really hurt either.
